java/amazon-kinesis-producer/src/main/java/com/amazonaws/services/kinesis/producer/KinesisProducer.java

```diff
@@ -601,6 +631,21 @@ protected KinesisProducer(File inPipe, File outPipe) {
         
         return f;
     }
+
+    @AllArgsConstructor
+    private class FutureTimeoutRunnableTask implements Runnable {
+        private long id;
+
+        @Override
+        public void run() {
+            SettableFutureTracker futureTracker = getFutureTracker(id);
+            totalFutureTimeouts.getAndIncrement();
+            SettableFuture<?> f = futureTracker.getFuture();
+            String message = "Message id " + id + " timeout out. Removing the submitted future from processing queue.";
+            f.setException(new FutureTimedOutException(message));
```

Comment:
Do we have any race condition here? when both timeout and normal code flow triggers at same time?

Reply:
yeah both need to call futures.remove(id); from the hashmap which will prevent only one of them to be executed.